### PR TITLE
BLAKE3: new port (v1.3.1)

### PR DESCRIPTION
**Pull request by srirangav**
#### Description

New port for b3sum from the BLAKE3 project

###### Type(s)

- [ ] bugfix
- [X] enhancement
- [ ] security fix

###### Tested on

macOS 11.6.4 20G417 arm64
Xcode 12.5.1 12E507

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [X] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

FYI, b3sum is written in rust and when I check the Portfile with port lint --nitpick, it gives me the same checksum warnings for rust / cargo dependencies that I see when linting other rust based ports (such as devel/gfold).  I am new to rust, so please let me know if there is a better way to implement this Portfile.

### CODE REVIEW

**Comment by herbygillot** [comment]
So I have not taken a look myself, but in this Portfile, `cargo build` is happening twice.

The first `cargo build` happens by default due to the `cargo` port group.  Your `build.pre_args-replace` affects the arguments for this invocation.

Then I see you manually do a 2nd `cargo build` in the `post-build`.

Since your destroot is only copying `b3sum`, then should that be the only thing you build? I think you can just set `build.dir b3sum` in the Portfile's main body to just focus on building the `b3sum` utility and just install that using `destroot`.  Doing it this way should eliminate the need for what you're doing in `post-build`.  Though I could be wrong.

File: security/BLAKE3/Portfile [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           cargo   1.0
+PortGroup           github  1.0
+
+github.setup        BLAKE3-team BLAKE3 1.3.1
+github.tarball_from archive
```
Comment (reneeotten) [inline security/BLAKE3/Portfile:8]: can we have the name not capitalized?

**Review by reneeotten — CHANGES_REQUESTED** [review]
thanks for the submission @srirangav, a few small comments. I am not really sure why this needs changes in `build.pre_args` and `post-build` phases and cannot use the defaults from the `cargo` PG. But I am not really versed in the cargo business, perhaps @herbygillot can look this over as well?

**Comment by herbygillot** [comment]
> thanks for the submission @srirangav, a few small comments. I am not really sure why this needs changes in `build.pre_args` and `post-build` phases and cannot use the defaults from the `cargo` PG. But I am not really versed in the cargo business, perhaps @herbygillot can look this over as well?

Yes, my comments are above.

File: security/BLAKE3/Portfile (hunk 1 above)
Comment (srirangav) [reply security/BLAKE3/Portfile]: Hi @reneeotten,

I have made this change in the Portfile.

Thanks,

-ranga

File: security/BLAKE3/Portfile [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+description         ${name} is a cryptographic hash function.
+long_description    ${name} is a cryptographic hash function that is \
+                    much faster than MD5, SHA-1, SHA-2, SHA-3, \
+                    and BLAKE2, secure, unlike MD5 and SHA-1, \
+                    secure against length extension, unlike SHA-2, \
+                    highly parallelizable, and capable of verified \
+                    streaming and incremental updates.
+
+categories          security
+installs_libs       yes
```
[the remark this answers is not in the document]
Comment (srirangav) [reply security/BLAKE3/Portfile]: Hi @reneeotten,

I have removed this from the Portfile.

Thanks,

-ranga

**Comment by srirangav** [comment]
> So I have not taken a look myself, but in this Portfile, you are doing a `cargo build` twice.
> 
> The first `cargo build` happens by default due to the `cargo` port group. Your `build.pre_args-replace` affects the arguments for this invocation.
> 
> Then I see you manually do a 2nd `cargo build` in the `post-build`.
> 
> Since your destroot is only copying `b3sum`, then should that be the only thing you build? I think you can just set `build.dir b3sum` in the Portfile's main body to just focus on building the `b3sum` utility and just install that using `destroot`. Doing it this way should eliminate the need for what you're doing in `post-build`. Though I could be wrong.

Hi @herbygillot,

I tried adding 'build.dir b3sum' but it wouldn't build properly, so I just changed the 'post-build' phase I had in the Portfile to the 'build' phase and it seems to work properly and only performs 1 compile.

Please let me know if this is okay.

Thanks,

-ranga

**Comment by herbygillot** [comment]
OK, so I had a chance to take a look at this.

I pushed a commit that fixes this Portfile to work as expected.

As mentioned several times already, issuing an explicit `cargo build` is not necessary as this is done automatically since this port is using the `cargo` portgroup.  Since the target that you're interested is in the `b3sum` directory, there were two changes needed:

1. Tell MacPorts to build in that directory using `build.dir ${worksrcpath}/b3sum`
2. Modify the path we install the binary from in destroot

Additionally, I'm not sure if you put the `cargo.crates` section together manually, but you don't need to.  There are some utilities in the ports tree to help you do that (`cargo2ports`), and I have one myself that you can also use: https://github.com/herbygillot/cargo2ports (basically just point it a `Cargo.lock` file and it will generate that section.  For this port, we are using the `Cargo.lock` in the `b3sum` directory).

**Comment by srirangav** [comment]
> I pushed a commit that fixes this Portfile to work as expected.

Thanks very much @herbygillot!  I haven't tried to put together a rust based Portfile before so I apologize for not quite getting this right.  I'll keep your changes in mind for the future.

Best,

-ranga

**Comment by herbygillot** [comment]
One other comment I had is that I've noticed you've set `nomaintainer` for all the ports you've submitted.

If you plan to return and update these ports as new versions of the come out, then you can set yourself as maintainer.  If you do so, you can also add `openmaintainer`, which indicates that other MacPorts project members can make minor updates to the port as well.

An example would look like:
```
maintainers         {@srirangav} openmaintainer
```

**Comment by herbygillot** [comment]
@srirangav I think for now go ahead and squash the two commits into your original commit so that you are the proper author, and the commit has the right commit description.  We'll merge afterwards.

**Comment by srirangav** [comment]
> maintainers         {@srirangav} openmaintainer

Hi @herbygillot,

I have squashed your commit with my original commit and have added myself as a maintainer as you suggested.  Hopefully the squashing worked (I am still somewhat new to git).

FYI, I did not include myself as a maintainer in the Portfiles I submitted recently because I thought that only members of the macports team could be maintainers.  In the future, when I submit a new Portfile, I will include myself as a maintainer.

Thanks very much for all your help!

Best,

-ranga

**Comment by herbygillot** [comment]
You can certainly open new pull requests to retroactively add yourself as maintainer on ports you have submitted in the past.

We'll leave this for @reneeotten to review then merge thereafter.  Thanks.

**Comment by herbygillot** [comment]
Merging this. Thanks @srirangav.

**Pull request merged**